File: solutions/python-refugee/1415_getHappyString.py
class Solution:
    def getHappyString(self, n: int, k: int) -> str:
        letters = ['a', 'b', 'c']
        happys = []

        def backtrack(curr, idx):
            if (len(curr) == n):
                happys.append(curr)
                return
            for i in range(len(letters)):
                if (i == idx):
                    continue
                backtrack(curr + letters[i], i)

        for i in range(len(letters)):
            backtrack(letters[i], i)

        return "" if k-1 >= len(happys) else happys[k-1]

File: solutions/1415_getHappyString.cc
// Feb. 18th, 2025 (daily)

class Solution {
    public:
        string getHappyString(int n, int k) {
            vector<char> letters = {'a','b','c'};
    
            vector<string> happys;
    
            for (int i = 0; i<3; ++i) {
                string happy(1, letters[i]);
                backtrack(happy, i, letters, n, happys);
            }
    
            return (k-1 >= happys.size()) ? "" : happys[k-1];
        }
    
        void backtrack(string& res, int& idx, const vector<char>& letters, const int& n, vector<string>& happys) {
            if (res.size() == n) {
                happys.push_back(res);
                return;
            }
    
            for (int i = 0; i<3; ++i) {
                if (i == idx) continue;
                res.push_back(letters[i]);
                backtrack(res, i, letters, n, happys);
                res.pop_back();
            }
    
        }
    };